HatchGenerator/Freecad_Hatch_Generator.py

```diff
@@ -0,0 +1,4355 @@
+            for prop in doc_obj.PropertiesList:
+                if prop not in excluded_props and not prop.startswith("Proxy"):
+                    setattr(new_obj, prop, getattr(doc_obj, prop))
+            FreeCAD.ActiveDocument.recompute()
+        except Exception as e:
+            FreeCAD.Console.PrintError(f"Copy failed: {str(e)}\n")
+
+    def duplicateHatch(self, doc_obj):
+        try:
+            new_name = FreeCAD.ActiveDocument.getUniqueObjectName("Hatch")
+            new_obj = makeCustomHatch(name=new_name)
+            excluded_props = ["Name", "Label", "ExpressionEngine"]
+            for prop in doc_obj.PropertiesList:
+                if prop not in excluded_props and not prop.startswith("Proxy"):
+                    setattr(new_obj, prop, getattr(doc_obj, prop))
+            FreeCAD.ActiveDocument.recompute()
+        except Exception as e:
+            FreeCAD.Console.PrintError(f"Duplicate failed: {str(e)}\n")
+
+    def removeHatch(self, doc_obj):
+        try:
+            FreeCAD.ActiveDocument.removeObject(doc_obj.Name)
+            FreeCAD.ActiveDocument.recompute()
+        except Exception as e:
+            FreeCAD.Console.PrintError(f"Remove failed: {str(e)}\n")
+
+    def doubleClicked(self, vobj):
+        runHatchGeneratorDialog()
+        return True
+
+    def attach(self, vobj):
+        self.ViewObject = vobj
+        self.Object = vobj.Object
+
+    def getDisplayModes(self, obj):
+        return ["Flat Lines", "Wireframe", "Shaded"]
+
+    def getDefaultDisplayMode(self):
+        return "Flat Lines"
+
+    def setDisplayMode(self, mode):
+        return mode
+
+    def onChanged(self, vobj, prop):
+        pass
+
+    def getIcon(self):
+        return """
+        /* XPM */
+        static char * hatch_xpm[] = {
+        "16 16 3 1",
+        " 	c None",
+        ".	c #000000",
+        "+	c #FFFFFF",
+        "                ",
+        " .  .  .  .  . ",
+        "                ",
+        "  .  .  .  .   ",
+        "                ",
+        " .  .  .  .  . ",
+        "                ",
+        "  .  .  .  .   ",
+        "                ",
+        " .  .  .  .  . ",
+        "                ",
+        "  .  .  .  .   ",
+        "                ",
+        " .  .  .  .  . ",
+        "                ",
+        "                "};
+        """
+
+    def __getstate__(self):
+        return {"Object": self.Object.Name}
+
+    def __setstate__(self, state):
+        if "Object" in state:
+            self.Object = FreeCAD.ActiveDocument.getObject(state["Object"])
+        return None
+
+def makeCustomHatch(name="CustomHatchFP"):
+    doc = FreeCAD.ActiveDocument
+    obj = doc.addObject("Part::FeaturePython", name)
+    CustomHatchFeature(obj)
+    CustomHatchViewProvider(obj.ViewObject)
+    return obj
+
+# ================================
+# GUI Dialog for creation
+# ================================
+class HatchGeneratorDialog(QtWidgets.QDialog):
```

Comment:
Please use a separate `ui` file for better maintainance (cf. `ObjectCreation/BoxCreator.FCMacro` or other macros for an example).